### PR TITLE
Make ARN Resource builder more expressive. Implement proper toString()

### PR DESCRIPTION
**Pull request by belugabehr**
## Motivation and Context

The modeling in `ArnResource` does not well match the spec. [1]

The toString() method of the `ArnResource` does not produce a proper (i.e., to spec) value.

The current implementation uses the phrase `resource` whereas the spec calls it a `resource-id`.  So, from the top-level `Arn` class, it's a bit awkward to access it: `myArn.getResource().getResource()`.

[1] https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html

## Modifications

* Deprecate `resource` attribute -> `resourceId`
* Update toString() to produce a properly formed output
* Allow for specifying a legacy '/' character to separate `resoure-type` and `resource-id`

## Testing
Added new unit tests

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: core/arns/src/main/java/software/amazon/awssdk/arns/ArnResource.java [hunk 1]
```diff
@@ -57,8 +59,16 @@ public Optional<String> resourceType() {
     /**
      * @return the entire resource as a string
      */
+    @Deprecated
```
Comment (zoewangg) [inline core/arns/src/main/java/software/amazon/awssdk/arns/ArnResource.java:62]: Instead of deprecating it, could we clarify it in the Javadoc?

File: core/arns/src/main/java/software/amazon/awssdk/arns/ArnResource.java [hunk 2]
```diff
@@ -103,22 +113,27 @@ public static ArnResource fromString(String resource) {
...
         }
+        if (splitter.equals('/')) {
+            builder.usingSlashSeperator();
+        }
 
         return builder.build();
     }
 
     @Override
     public String toString() {
```
Comment (zoewangg) [inline core/arns/src/main/java/software/amazon/awssdk/arns/ArnResource.java:129]: This may potentially break customers who rely on existing behavior. Can we introduce a new method, `toStringPretty` or `toStringFormatted`?

File: core/arns/src/main/java/software/amazon/awssdk/arns/ArnResource.java [hunk 3]
```diff
@@ -184,6 +208,15 @@ public interface Builder extends CopyableBuilder<ArnResource.Builder, ArnResourc
...
         Builder qualifier(String qualifier);
 
+        /**
+         * For legacy AWS Arns not following the resourceType:resource:qualifier pattern,
+         * the qualifier field will contain everything after the first two sections separated
+         * by either a slash character '"/".
+         *
+         * @return Returns a reference to this builder
+         */
+        Builder usingSlashSeperator();
```
Comment (zoewangg) [inline core/arns/src/main/java/software/amazon/awssdk/arns/ArnResource.java:218]: Should we just take a character as separator? 
```
Builder seperator(Character char);
```

**Comment by belugabehr** [comment]
@debora-ito Other priorities keeping me from finishing - but my intent is to see it through.  Thanks for the patience.

**Comment by cenedhryn** [comment]
@belugabehr Any updates? Saw you updated that other PR :-)

**Comment by belugabehr** [comment]
@cenedhryn Hello - better late than never.  I made the the requested changes.  However, I notice a small bug that is also fixed in this PR.

The _qualifier_ of an ARN is always preceded by a colon ':'.  However, the current code allows it to be a colon or a slash - this is incorrect.


https://github.com/aws/aws-sdk-java-v2/blob/c2db6affbf7e33066072b0b366a18c8c4db57e3c/core/arns/src/main/java/software/amazon/awssdk/arns/ArnResource.java#L104

File: core/arns/src/test/java/software/amazon/awssdk/arns/ArnTest.java [hunk 4]
```diff
@@ -110,8 +111,7 @@ public void arnWithResourceTypeAndResourceAndQualifier_SlashSplitter_ParsesCorre
         assertThat(arn.service()).isEqualTo("s3");
         assertThat(arn.region()).hasValue("us-east-1");
         assertThat(arn.resourceAsString()).isEqualTo("bucket/foobar/1");
-        verifyArnResource(arn.resource());
-        assertThat(arn.resource().qualifier().get()).isEqualTo("1");
+        assertThat(arn.resource().qualifier()).isEqualTo(Optional.empty());
```
Comment (belugabehr) [inline core/arns/src/test/java/software/amazon/awssdk/arns/ArnTest.java:114]: As seen here, this is a valid S3 path, yet it is being treated as a qualifier.

File: core/arns/src/test/java/software/amazon/awssdk/arns/ArnTest.java [hunk 5]
```diff
@@ -31,7 +33,6 @@ public void arnWithBasicResource_ParsesCorrectly() {
         assertThat(arn.region()).hasValue("us-east-1");
         assertThat(arn.accountId()).hasValue("12345678910");
         assertThat(arn.resourceAsString()).isEqualTo("myresource");
-        System.out.println(arn.resource());
```
Comment (belugabehr) [inline core/arns/src/test/java/software/amazon/awssdk/arns/ArnTest.java:34]: General cleanup

File: core/arns/src/test/java/software/amazon/awssdk/arns/ArnResourceTest.java [hunk 6]
```diff
@@ -64,4 +66,72 @@ public void hashCodeEquals_minimalProperties() {
         assertThat(arnResource.equals(anotherResource)).isTrue();
         assertThat(arnResource.hashCode()).isEqualTo(anotherResource.hashCode());
     }
+
+    @Test
+    public void toString_resourceIdOnly() {
+        ArnResource arnResource = ArnResource.builder().resourceId("resource").build();
+        assertThat("null:resource:null").isEqualTo(arnResource.toString());
```
Comment (belugabehr) [inline core/arns/src/test/java/software/amazon/awssdk/arns/ArnResourceTest.java:73]: Add unit test to demonstrate the unexpected 'null' behavior here of the toString
